src/mime.cr

```diff
@@ -0,0 +1,200 @@
+  # matched case-insensitive. Returns *default* if *extension* is not registered.
+  def self.from_filename(filename : String, default) : String
+    from_extension(File.extname(filename), default)
+  end
+
+  # Looks up the MIME type associated with the extension in *filename*.
+  #
+  # A case sensitive search is tried first, if this yields no result, it is
+  # matched case-insensitive. Raises `KeyError` if extension is not registered.
+  def self.from_filename(filename : String) : String
+    from_extension(File.extname(filename))
+  end
+
+  # Looks up the MIME type associated with the extension in *filename*.
+  #
+  # A case sensitive search is tried first, if this yields no result, it is
+  # matched case-insensitive. Returns `nil` if extension is not registered.
+  def self.from_filename?(filename : String) : String?
+    from_extension?(File.extname(filename))
+  end
+
+  # Looks up the MIME type associated with the extension in *filename*.
+  #
+  # A case sensitive search is tried first, if this yields no result, it is
+  # matched case-insensitive. Runs the fiven block if extension is not registered.
+  def self.from_filename(filename : String, &block)
+    from_extension(File.extname(filename)) { |extension| yield extension }
+  end
+
+  # Register *type* for *extension*.
+  #
+  # *extension* must start with a dot (`.`) and must not contain any null bytes.
+  def self.register(extension : String, type : String) : Nil
+    raise ArgumentError.new("Extension does not start with a dot: #{extension.inspect}") unless extension.starts_with?('.')
+    extension = extension.check_no_null_byte
+
+    initialize_types
+
+    mediatype = parse_media_type(type) || raise Error.new "Invalid media type: #{type}"
+
+    @@types[extension] = type
+    @@types_lower[extension.downcase] = type
+
+    if mediatype
+      array = @@extensions[mediatype] ||= [] of String
+      array << extension
+    end
+
+    nil
+  end
+
+  # Returns all extensions registered for *type*.
+  def self.extensions(type : String) : Array(String)
+    @@extensions.fetch(type) { [] of String }
+  end
+
+  # tspecial as defined by RFC 1521 and RFC 2045
+  private TSPECIAL_CHARACTERS = {'(', ')', '<', '>', '@', ',', ';', ':', '\\', '"', '/', '[', ']', '?', '='}
+
+  # :nodoc:
+  def self.parse_media_type(type : String) : String?
+    reader = Char::Reader.new(type)
+
+    sub_type_start = -1
+    while reader.has_next?
+      case char = reader.current_char
+      when ';'
+        break
+      when '/'
+        return nil if sub_type_start > -1
+        sub_type_start = reader.pos
+        reader.next_char
+      else
+        if TSPECIAL_CHARACTERS.includes?(char) || 0x20 > char.ord > 0x7F
+          return nil
+        end
+
+        reader.next_char
+      end
+    end
+
+    if reader.pos == 0
+      return nil
+    end
+
+    type.byte_slice(0, reader.pos).strip.downcase
+  end
+
+  # :nodoc:
+  def self.load_mime_database(io : IO) : Nil
```

Comment:
Either :
- make it public —so we can load whatever `mime.types` file we want, which would be quite useful (especially on Windows);
- make it private —and have Crystal::System::MIME merely tell where the `mime.types` files are located;
- move it to `Crystal::System::MIME` that will return/yield mime types that `MIME` can register (compatible with Windows' registry, no circular dependency).

Reply:
Yes, it should be public.